projects/core/src/lib/bookmark/bookmark.service.ts

```diff
@@ -1,224 +1,210 @@
+    // compressed
+    const binaryFragments = new BinaryBookmarkFragments({ version: BookmarkService.BINARY_FRAGMENTS_VERSION_1 });
 
-      if (pendingFragment !== undefined) {
-          info.observable.next(info.deserializer(pendingFragment));
-          this.pendingFragments.delete(descriptor.identifier);
-      }
+    const fragmentById = new Map<string, [BookmarkFragmentDescriptor, BookmarkFragmentValueObservable]>();
+    for (const [ key, fragment ] of this.fragments) {
+      fragmentById.set(key.identifier, [ key, fragment ]);
+    }
 
-      this.fragments.set(descriptor, info);
-      return info.observable;
-  }
+    for(const id of [...fragmentById.keys()].sort()) {
+      const fragment = fragmentById.get(id) as [BookmarkFragmentDescriptor, BookmarkFragmentValueObservable];
+      const descriptor = fragment[0];
+      const value = fragment[1].value;
 
-  public updateFragment<T extends Message<T>>(descriptor: BookmarkFragmentProtoDescriptor<T>, message: T): void;
-  public updateFragment(descriptor: BookmarkFragmentStringDescriptor, message: string): void;
-  public updateFragment(descriptor: AnyDescriptor, message: any) {
-      if (!this.fragments.has(descriptor)) {
-          this.registerFragment$(descriptor as any);
+      if (!value || value === '') {
+        continue;
       }
 
-      const fragment = this.fragments.get(descriptor);
-      if (fragment === undefined) {
-          return;
+      if (descriptor.identifier === '') {
+        // This will be first when sorted by identifier
+        // Value is location, no need for URL encoding
+        output += '@' + value;
+      } else if (descriptor.type === 'string') {
+        if (value.includes('!')) {
+          throw new Error('String bookmark fragment value may not contain \'!\' separator');
+        }
+        // Identifier does not need to be URL encoded, see regexp check in registerFragment$()
+        output += `!${descriptor.identifier}:` + encodeURIComponent(value);
+      } else if (descriptor.type === 'binary') {
+        const bytes = descriptor.serialize(value) as Uint8Array;
+        if (bytes.length > 0) {
+          binaryFragments.fragments.push(new BookmarkFragment({
+            identifier: descriptor.identifier,
+            bytes,
+          }));
+        }
       }
+    }
 
-      fragment.observable.next(message);
-      this._serializeBookmark();
-  }
-
-  private _serializeBookmark() {
-      const components: Map<string, string> = new Map();
-      for (const [ key, component ] of this.fragments) {
-          const serialized = component.serializer(component.observable.value);
-          if (serialized === undefined) {
-              continue;
-          }
-          if (serialized.length === 0) {
-              continue;
-          }
+    if (binaryFragments.fragments.length > 0) {
+      const protobuf = binaryFragments.toBinary();
+      const deflater = new Deflater({ format: 'raw', level: 9 });
+      deflater.append(protobuf);
```

Comment:
Missed this: `deflate.append` may also return blocks, though unlikely at this scale. (This might also explain why the inflate works unreliably - it probably just desyncs the compressed stream.)

Reply:
Good catch, I think just using the `deflate()` function is better, no need to work with chunks.